rllib/core/rl_trainer/torch/torch_rl_trainer.py

```diff
@@ -1,10 +1,185 @@
 import logging
-from ray.rllib.core.rl_trainer.rl_trainer import RLTrainer
+from typing import (
+    Any,
+    Mapping,
+    Union,
+    Type,
+    Sequence,
+    Hashable,
+    Optional,
+    Callable,
+)
+import torch
 
-logger = logging.getLogger(__name__)
+from ray.train.torch.train_loop_utils import _TorchAccelerator
+
+from ray.rllib.core.rl_module.rl_module import RLModule, ModuleID
+from ray.rllib.core.rl_trainer.rl_trainer import (
+    RLTrainer,
+    MultiAgentRLModule,
+    ParamOptimizerPairs,
+    Optimizer,
+    ParamType,
+    ParamDictType,
+)
+from ray.rllib.core.rl_module.torch.torch_rl_module import DDPRLModuleWrapper
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.typing import TensorType
+from ray.rllib.utils.nested_dict import NestedDict
 
-# TODO: Implement this
+logger = logging.getLogger(__name__)
 
 
 class TorchRLTrainer(RLTrainer):
-    pass
+    def __init__(
+        self,
+        module_class: Union[Type[RLModule], Type[MultiAgentRLModule]],
+        module_kwargs: Mapping[str, Any],
+        scaling_config: Mapping[str, Any],
+        optimizer_config: Mapping[str, Any],
+        distributed: bool = False,
+        in_test: bool = False,
+    ):
+        super().__init__(
+            module_class=module_class,
+            module_kwargs=module_kwargs,
+            scaling_config=scaling_config,
+            optimizer_config=optimizer_config,
+            distributed=distributed,
+            in_test=in_test,
+        )
+
+        self._world_size = scaling_config.get("num_workers", 1)
+        self._use_gpu = scaling_config.get("use_gpu", False)
+        # These attributes are set in the `build` method.
+        self._device = None
+
+    @property
+    @override(RLTrainer)
+    def module(self) -> MultiAgentRLModule:
+        return self._module
+
+    @override(RLTrainer)
+    def configure_optimizers(self) -> ParamOptimizerPairs:
+        lr = self.optimizer_config.get("lr", 1e-3)
+        return [
+            (
+                self.get_parameters(self._module[key]),
+                torch.optim.Adam(self.get_parameters(self._module[key]), lr=lr),
+            )
+            for key in self._module.keys()
+        ]
+
+    @override(RLTrainer)
+    def compute_gradients(
+        self, loss: Union[TensorType, Mapping[str, Any]]
+    ) -> ParamDictType:
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
+        loss[self.TOTAL_LOSS_KEY].backward()
+        grads = {pid: p.grad for pid, p in self._params.items()}
+
+        return grads
+
+    @override(RLTrainer)
+    def apply_gradients(self, gradients: ParamDictType) -> None:
+
+        # make sure the parameters do not carry gradients on their own
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
```

Comment:
Why `set_to_none`? Is it faster? If so, can you add a one-line comment here?

Reply:
yes.

Reply:
added